x-pack/plugins/cloud_security_posture/public/components/fleet_extensions/package_policy_input_var_field.tsx

```diff
@@ -0,0 +1,476 @@
+  return (
+    <>
+      <EuiFlexGroup alignItems="flexEnd" gutterSize="xs">
+        <EuiFlexItem grow={false} aria-label={fieldLabel}>
+          {fieldLabel}
+        </EuiFlexItem>
+        <EuiFlexItem grow={false}>
+          <EuiToolTip
+            content={
+              <FormattedMessage
+                id="xpack.csp.createPackagePolicy.stepConfigure.secretLearnMorePopoverContent"
+                defaultMessage="This value is a secret. After you save this integration policy, you won't be able to view the value again."
+              />
+            }
+          >
+            <EuiIcon aria-label="Secret value" type="questionInCircle" color="subdued" />
+          </EuiToolTip>
+        </EuiFlexItem>
+      </EuiFlexGroup>
+
+      <EuiSpacer size="s" />
+    </>
+  );
+};
+
+function SecretInputField({
+  varDef,
+  value,
+  onChange,
+  frozen,
+  packageName,
+  packageType,
+  datastreams = [],
+  isEditPage,
+  isInvalid,
+  fieldLabel,
+  fieldTestSelector,
+  setIsDirty,
+  isDirty,
+}: InputComponentProps) {
+  const [isReplacing, setIsReplacing] = useState(isEditPage && !value);
+  const valueOnFirstRender = useRef(value);
+
+  const hasExistingValue = !!valueOnFirstRender.current;
+  const lowercaseTitle = varDef.title?.toLowerCase();
+  const showInactiveReplaceUi = isEditPage && !isReplacing && hasExistingValue;
+  const valueIsSecretRef = value && value?.isSecretRef;
+
+  const inputComponent = getInputComponent({
+    varDef,
+    value: isReplacing && valueIsSecretRef ? '' : value,
+    onChange,
+    frozen,
+    packageName,
+    packageType,
+    datastreams,
+    isEditPage,
+    isInvalid,
+    fieldLabel,
+    fieldTestSelector,
+    isDirty,
+    setIsDirty,
+  });
+
+  // If there's no value for this secret, display the input as its "brand new" creation state
+  // instead of the "replace" state
+  if (!hasExistingValue) {
+    return inputComponent;
+  }
+
+  if (showInactiveReplaceUi) {
+    return (
+      <>
+        <EuiText size="s" color="subdued">
+          <FormattedMessage
+            id="xpack.csp.editPackagePolicy.stepConfigure.fieldSecretValueSet"
+            defaultMessage="The saved {varName} is hidden. You can only replace the {varName}."
+            values={{
+              varName: lowercaseTitle,
+            }}
+          />
+        </EuiText>
+        <EuiSpacer size="s" />
+        <EuiButtonEmpty
+          onClick={() => setIsReplacing(true)}
+          color="primary"
+          iconType="refresh"
+          iconSide="left"
+          size="xs"
+        >
+          <FormattedMessage
+            id="xpack.csp.editPackagePolicy.stepConfigure.fieldSecretValueSetEditButton"
+            defaultMessage="Replace {varName}"
+            values={{
+              varName: lowercaseTitle,
+            }}
+          />
+        </EuiButtonEmpty>
+      </>
+    );
+  }
+
+  if (isReplacing) {
+    const cancelButton = (
```

Comment:
Apologies,  I should have been more informative.  The reason why I mentioned this is because we should try to avoid unnecessary re-renders.  The cancellation is acting like a mini react component that is re-rendering every time it is instantiated.